jobbergate-core/jobbergate_core/auth/core.py

```diff
@@ -0,0 +1,316 @@
+    .. _httpx: https://www.python-httpx.org/
+
+    Arguments:
+        cache_directory (Path): Directory to be used for the caching tokens.
+        login_domain (str): Domain used for the login.
+        login_audience (str): Audience of the login.
+        login_client_id (str): Client ID used for login.
+        tokens (Dict[TokenType, Token]): Dictionary holding the tokens needed for authentication.
+
+    Note:
+        Consult the values above with your system administrator.
+
+    Note:
+        This class can interoperate with the tokens generated by the ``jobbergate-cli`` package, as long
+        as they are stored in the same cache directory.
+
+    Examples:
+
+        The following example shows how to use the JobbergateAuth class to authenticate a request:
+
+        >>> from pathlib import Path
+        >>> import requests
+        >>> from jobbergate_core import JobbergateAuth
+        >>> jobbergate_auth = JobbergateAuth(
+        ...     cache_directory=Path("."),
+        ...     login_domain="http://keycloak.local:8080/realms/jobbergate-local",
+        ...     login_audience="https://local.omnivector.solutions",
+        ...     login_client_id="cli",
+        ... )
+        >>> jobbergate_base_url = "http://localhost:8000/jobbergate"
+        >>> jobbergate_auth.acquire_tokens()
+        Login Here: http://keycloak.local:8080/realms/jobbergate-local/device?user_code=LMVJ-XOLG
+        >>> response = requests.get(
+        ...     f"{jobbergate_base_url}/applications",
+        ...     auth=jobbergate_auth # this is the important part
+        )
+        >>> response.raise_for_status()
+        >>> print(f"response = {response.json()}")
+    """
+
+    cache_directory: Path
+    login_domain: str
+    login_audience: str
+    login_client_id: str
+    tokens: Dict[TokenType, Token] = field(default_factory=dict)
+
+    def __call__(self, request):
+        """
+        This internal method allows the integration with the ``requests`` library.
+
+        It is called automatically when the instance is passed to the ``auth`` parameter,
+        see the examples in the class docstring.
+
+        It adds the ``Authorization`` header to the request with the access token.
+        """
+        logger.debug("Authenticating request")
+
+        self.acquire_tokens()
+        access_token = self.check_credentials(token_type=TokenType.ACCESS)
+
+        request.headers["Authorization"] = f"Bearer {access_token.content}"
+        return request
+
+    def acquire_tokens(self):
+        """
+        High-level method to acquire a valid access token.
+
+        This method will attempt, in order:
+
+        * Load the tokens from the cache directory (see :meth:`JobbergateAuth.load_from_cache`)
+        * If the access token is unavailable or expired, refresh both tokens
+          using the refresh token (see :meth:`JobbergateAuth.refresh_tokens`)
+        * If the refresh token is unavailable or expired, login to fetch both tokens
```

Comment:
`Fetch` isn't the most appropriate word IMO.
```suggestion
        * If the refresh token is unavailable or expired, login to generate both tokens
```